baseservice/base_service.py

```diff
@@ -0,0 +1,122 @@
+import logging
+import signal
+import threading
+from abc import ABCMeta, abstractmethod
+from enum import Enum, unique
+from typing import Optional
+
+from baseservice.utils import Event
+
+
+@unique
+class ServiceState(Enum):
+    STARTING = 'STARTING'
+    STOPPING = 'STOPPING'
+    STOPPED = 'STOPPED'
+
+
+class BaseService(metaclass=ABCMeta):
+    """
+    this class is the base class for all services
+    """
+
+    def __init__(self, *, name: str = None, should_stop_on_signal=True):
+        """
+
+        :param name: the name of this service. if None, the name of the type will be used
+        :param should_stop_on_signal: if True, the service will try to register SIGTERM and SIGINT on stop method.
+        """
+        self._should_stop_on_signal: bool = should_stop_on_signal
+        self._cancellation_token: threading.Event = threading.Event()
+        self._cancellation_token.set()  # service starts as not_running
+        self._logger = logging.getLogger(__name__)
+        self._state_event: Event[ServiceState] = Event()
+        if not name:
+            name = type(self).__name__
+
+        self._name = name
+
+    @property
+    def name(self) -> str:
+        return self._name
+
+    @property
+    def is_alive(self) -> bool:
+        """
+        this property is True, if the service is running.
+        """
+        return self._is_alive()
+
+    @property
+    def state_event(self) -> Event[ServiceState]:
+        """
+        this is an Event, that can be used to register on server state changes
+        """
+        return self._state_event
+
+    def _is_alive(self) -> bool:
+        """
+        this may be overridden by child classes to change the 'is_alive' behaviour
+        """
+        return not self._cancellation_token.is_set()
+
+    def start(self):
+        """
+        starts the service, and blocks until 'stop' is called
+        """
+        self._cancellation_token.clear()
+        if self._should_stop_on_signal:
+            self._register_signals()
+        server_exception = None
+        try:
+            self._state_event.fire(ServiceState.STARTING)
+            self._logger.info(f"Starting {self.__name__}")
+            self._prepare_service()
+            self._run_service(cancellation_token=self._cancellation_token)
```

Comment:
Why do we pass the cancelation token as an argument? The implementation already has access to the property

Reply:
good point.
i did that, so it would make it explicit to the implementer that he should use it. it's kind of a documentation...
but it's really not necessary... 
is is that annoying?